### ISSUE TITLE
[FEAT] Display delta of new threads/microblogs since last visit

### ISSUE THREAD

**Issue by kotking** [aclist/kbin-kes#217]
I'm always frustrated when I need to look up Microblog section to be sure if something new appeared and as moderator to delete nsfw posts that listening to # brought up. Unless Kbin adds pre approval of posts to appear from other instances, moderation team taking quick glance if new posts appeared and if they sfw or not.

**Describe the solution you'd like**
Have toggle in KES to show microblog near threads, maybe even as switch between microblog and threads posts. 

**Describe alternatives you've considered**
Not bothering to look on mobile, or scroll all threads till next page available and only before side description will show microblog section.

**Additional context**

**Comment by aclist** [comment]
Thanks for the suggestion. I believe I understand the basic concept behind this proposal, but since I do not moderate any magazines that have active microblog posts alongside threads, I'm not sure of the practical implications of what you are asking for.

As I understand it, you want to have a unified view that merges both microblog posts and regular thread posts into a single thread view so that you can moderate the whole queue without having to switch between them. Is this correct?

It would help if you could post some screenshots of the current workflow, as well as some mockups (if possible) of what you would like it to be. In the meantime, I will try to simulate this by creating some testing magazines and posts and taking a look at the moderation queue.

**Comment by kotking** [comment]
I will admit being a bit... slow on this one.
Still [Here what I did](https://imgur.com/keNKPf1) .
Here[ how it's on PC](https://i.imgur.com/FvMRmAM.png) .
I was scrolling down to switch to microblog section, instead of tapping on icon on top.
Still on PC you see amount of posts as microblogs posts are all separate most of the time and grabbed randomly, if there is nsfw one I will detect by looking if the amount of posts increased as it mostly buggy and rarely pick up lately.  
So having option to see amount of threads/microblog posts and easy switch between them, that's what I have in mind. 
So right now you can switch either by scrolling down to description and choose microblog section or by clicking on Kbin icon, I didn't think of clicking on the icon. Thank you for giving attention to this annoyance I have.

**Comment by aclist** [comment]
Thanks for the additional videos. I took some time to investigate and test this on some instances. 

Basically, as you found, if you navigate to https://kbin.social/mod (via the list icon at the top), it will take you to a new moderation page where you can then navigate to either Threads or Microblogs from the navbar.

I don't see an easy way of merging the content of these two because they don't exist on the same page; when you click on Microblogs, for example, it has to load in this content remotely and fetch those posts. In addition, microblog style posts have a different threading and comment structure, since they come in from a variety of sources. If we were to attempt to forcibly merge the content of these two, we would have to first poll the threads and microblogs ahead of time and then push them all into one page. The problem with this is that you are effectively doubling the time it takes to load, since it's not just about showing the other content--we have to fetch it, too. There is also the question of how you would merge these different types of content, since they don't have the same threading structure.

So in practice, the moderation page would take twice as long to load if we opted to load in this content in the background.

I don't think merging is the correct approach here, but I could see how some fast navigation button might help, if you feel that going to https://kbin.social/mod is still not enough.

For example, we could add two additional links to the list icon next to your profile (three dots and three lines).

BEFORE:
```
Subscribed
Favorited
Moderated
All
```

AFTER:
```
Subscribed
Favorited
Moderated (Threads)
Moderated (Microblogs)
All
```

In essence, this would save you one step, as you'd be able to go to, e.g., Microblog moderation directly from the top navbar. I don't know if this is more convenient for you, but it's the best I could think of given the restrictions above.

**Comment by kotking** [comment]
![IMG_20231025_165854](https://github.com/aclist/kbin-kes/assets/18371182/aac9fb3c-9594-43f3-978f-25e2c1c66bd0)
![IMG_20231025_165941](https://github.com/aclist/kbin-kes/assets/18371182/6acada79-3859-43d3-894b-397f1f31dc01)
  
I wasn't aiming on big change, but from screenshots provided you can see that neither landscape mode nor vertical mode shows amount of threads and posts. It will show if I switch to "pc mode" in mobile browser, so my hindrance is not having displayed as on pc but hidden under menu. So my gripe was firstly that you need to be in magazine to see numbers and can't see them from moderation tab( depending on magazine/s  size and amount of moderation required could be useless or useful) and second that even when I am on magazine I want to check, it is hidden in drop-down menu rather to be seen on top bar as on pc. So it's more of a I need to track numbers to see if there was activity and I need intervention (as you can't control the Microblog section, only manage what it can grab for now) than merging both threads and Microblogs to check on activity which is how Lemmy works (you can directly post to Lemmy community by addressing community in body of the message). So if you can fit in little tracker to replace for example Logo with T(200+) M(499+) that is would be all I ask really. Again thank you for dedicating time on this ( I Hope) easy request.

**Comment by aclist** [comment]
I understand what you are saying. I noticed that on mobile view, even if you are not a moderator, there is no indication of what magazine you are on or of the thread/microblog count. It seems all of this information gets truncated from the navbar. So I can see how your suggestion could be useful to general users and not just moderators. Below is a proposed mockup. Thoughts? This is easy to do and is minimally invasive.

![1698252898](https://github.com/aclist/kbin-kes/assets/92275929/a769a7c7-605f-4b93-a0d3-74435a3909fe)

**Comment by kotking** [comment]
Yeah I envisioned something along these lines. It could be improved upon, but I think mockup nail it in.

**Comment by aclist** [comment]
The mockup is merely to align expectations and set a baseline. I want to understand what the minimum required functionality would be, here:

Based on your earlier comments, it sounds like adding an incremental count is useful for small magazines (e.g. those with under 1,000 threads), since if you are a moderator or reader of the magazine, you will immediately notice if a magazine with 20 threads suddenly has 30, so you can tell the count went up. Whereas if the magazine has 1,000+ threads, the exact number stops being so useful, because it's already exceeded a large threshold. Is this more or less correct?

- Requirement 1: Round/truncate numbers larger than XXXX significant digits (e.g., 1,059 -> 1,000+. Prevent magazines with 1,000,000 posts from showing the literal number)
- Requirement 2: Show the real number if the post count is up to 3 significant digits (<= 999)
- Requirement 3: Add a label listing the current magazine being browsed


Questions: 
- Should tapping this element open a menu to jump to theads/microblogs?
- What else would be desirable here?

**Comment by kotking** [comment]
Yeah I can see problem for magazines/communities that has lot of activity...  
What about counter? Think this will be good solution and help for everyone and will be useful on PC version as well. just simple start count/reset count with same limit in 1000 posts. 
I could easily see when all problems with federation fixed how microblog section can go over 1000 while threads would still in 300+/-, mostly due nature how kbin grabs microblog posts.  
 If ability to jump between threads/microblogs is easy implement I wouldn't say no, but then it would need more space (big thumbs) so would have to compromise somewhere.  
From what I can see if counter is easy to implement it would make useful to check whats new on your magazines/communities (maybe options for daily/weekly/monthly as well) if not then just having for mobile version community=threads/microblogs with cap 1000 is good enough. Having ability to tap would mostly require to replace logo or would be useful in landscape but be obscured in portrait mode.

**Comment by aclist** [comment]
If I understood you correctly, you are proposing a method of marking the latest thread count as "read", storing the total number of last seen threads, then updating this counter if there are new threads at a later date (i.e., last seen thread count != current thread count).

In other words, the label `New threads since last visit: NNN` would be active if the number since last reset is greater than zero.

In practical terms:

- You visit a magazine for the first time
- The number of "new" threads/microblogs is listed as 100/200 respectively (at this point in time, new threads == total number of threads)
- You click Reset Count
- The values 100 and 200 are stored as the last seen count
- Next time you visit the magazine, if the number is greater than 100/200, the new thread count is updated and Reset Count button is available again
- If the number is not greater than 100/200, the new count label is not updated

This is a bit overengineered in my opinion, but I think you are looking for a way of noticing if there was a change in activity. At the most rudimentary, we need a way of knowing if there is new content since the last visit. Correct?

As far as tappability is concerned, have a look at the option `General > Subs Omnibar` and enable the setting `Add tappable element to screen`. According to testing, this should be easy to tap on mobile at its current size. In order to not occlude the navigation area further, the proposed thread count element would have a similar size and configuration.

**Comment by aclist** [comment]
@kotking Hello, there has not been activity here for some time. Do you have any final thoughts on the design of this add-on before I implement it?

**Comment by kotking** [comment]
A bit much to process, but yeah I was thinking along those lines. It could be made more feature rich or the most simplest way.
I would say the most simplest way works for me personally.
Go to x Magazine 
click start counting
when I visit same magazine it will scan if total amount of threads and microblogs changed from when it started tracking it had example 1000 threads when I clicked "start counting" > I came back it has 1030> Tracker shows +30> goes until either 100 or 999> click reset it takes 1030 as new starting point.  
It could be made complex with starting date/end date in settings and others QoL ideas, but that I leave for your interpretation.
Again thank you for work and looking forward to implementation.

**Comment by aclist** [comment]
This feature is now available in the [testing branch](https://github.com/aclist/kbin-kes/raw/testing/kes.user.js). The add-on is called `Show thread/microblog delta since last visit` and is available in the `Threads` page.

I have tried to consolidate the requirements above into something more automatic:

- The user can choose between:
    - Permanently displaying metadata in a bar above the page
    - Only displaying this bar if either new threads OR new microblogs are available. Mobile users who don't have thread/microblog counts in the navbar would probably prefer to set this to ON so that the bar is permanent. It will display the following information:

`magazine name@instance name | Threads (count) | Microblogs (count)`

- The user can customize the foreground and background color of this bar.

- Each time you visit a particular magazine on a particular instance, the current thread count and microblog count are saved into storage. The next time you visit that magazine, if the number of threads or microblogs has changed, that number is shown as a delta in the bar:

`magazine name | Threads (count) Δ number of new posts  | Microblogs (count) Δ number of new posts`

For instance,

`kbinmeta | Threads (145) Δ 2 | Microblogs (16)`

This indicates that there are `2` new threads, for a total of `145`, on `kbinmeta` since your last visit, up from `143` threads. The new total count is stored. Next time you visit the page, the bar will read:

`kbinmeta | Threads (145) | Microblogs (16)`

This indicates that the total thread count is `145` and there are no new threads.

When the bar is not in permanent mode, it will be hidden unless there are new threads/microblogs. Whether you choose to permanently enable the bar or have it appear on-demand, the number of new threads/microblogs will always be available for any magazine.

Note that the testing branch may introduce breaking changes from time to time, so after testing this feature, you can go back to the [stable branch](https://github.com/aclist/kbin-kes/raw/main/kes.user.js).

**Comment by aclist** [comment]
Resolved in 3.2.0
